Proofs: locals and memory identity proofs (#214)
kframework/wasm-semantics: data.md

Wasm memories are byte arrays, sized in pages of 65536 bytes, initialized to be all zero bytes.
To avoid storing many zeros in what may be sparse memory, we implement memory as maps, and store only non-zero bytes.
The maps store integers, but maintains the invariant that each stored integer is between 1 and 255, inclusive.
`#setRange(BM, START, VAL, WIDTH)` writes the integer `I` to memory as bytes (little-endian), starting at index `N`.

```k
    syntax Map ::= #setRange(Map, Int, Int, Int) [function]
 // -------------------------------------------------------
    rule #setRange(BM, IDX,   _, N) => BM
      requires N  ==Int 0
    rule #setRange(BM, IDX, VAL, N) => #setRange(#set(BM, IDX, VAL modInt 256), IDX +Int 1, VAL /Int 256, N -Int 1)

>       requires N =/=Int 0

Reviewer: Let's use `N >Int 0` and `notBool (N >Int 0)` for side conditions, to (1) ensure completeness and (2) ensure termination.

Author: Done

```

`#getRange(BM, START, WIDTH)` reads off `WIDTH` elements from `BM` beginning at position `START`, and converts it into an unsigned integer.
The function interprets the range of bytes as little-endian.

```k
    syntax Int ::= #getRange ( Map , Int , Int ) [function]
 // -------------------------------------------------------
    rule #getRange(BM, START, WIDTH) => 0
      requires WIDTH ==Int 0
    rule #getRange(BM, START, WIDTH) => #get(BM, START) +Int (#getRange(BM, START +Int 1, WIDTH -Int 1) *Int 256)

>       requires WIDTH >Int 0

Reviewer: Let's use `WIDTH >Int 0` and `notBool (WIDTH >Int 0)` for side conditions, which (1) ensures totalness and (2) ensures termination.

Author: Done

```

`#setRange(BM, START, VAL, WIDTH)` writes the integer `I` to memory as bytes (little-endian), starting at index `N`.

```k
    syntax ByteMap ::= #setRange(ByteMap, Int, Int, Int) [function]
 // ---------------------------------------------------------------
    rule #setRange(BM, IDX,   _, WIDTH) => BM
      requires notBool (WIDTH >Int 0)
    rule #setRange(BM, IDX, VAL, WIDTH) => #setRange(#set(BM, IDX, VAL modInt 256), IDX +Int 1, VAL /Int 256, WIDTH -Int 1)
      requires          WIDTH >Int 0
```

`#getRange(BM, START, WIDTH)` reads off `WIDTH` elements from `BM` beginning at position `START`, and converts it into an unsigned integer.
The function interprets the range of bytes as little-endian.

```k
    syntax Int ::= #getRange (ByteMap, Int , Int) [function]
 // --------------------------------------------------------
    rule #getRange(BM, START, WIDTH) => 0
      requires notBool (WIDTH >Int 0)
    rule #getRange(BM, START, WIDTH) => #get(BM, START) +Int (#getRange(BM, START +Int 1, WIDTH -Int 1) *Int 256)
      requires          WIDTH >Int 0
```

`#get` looks up a key in a map, defaulting to 0 if the map does not contain the key.

> `#set` sets a key in a map, defaulting to 0 if the value is 0.

Reviewer: ```suggestion
`#set` sets a key in a map, unsetting that key if the value is 0.
```

      requires N >Int 0
```

`#getRange(BM, START, WIDTH)` reads off `WIDTH` elements from `BM` beginning at position `START`, and converts it into an unsigned integer.
The function interprets the range of bytes as little-endian.

```k
    syntax Int ::= #getRange ( Map , Int , Int ) [function]
 // -------------------------------------------------------
    rule #getRange(BM, START, WIDTH) => 0
      requires WIDTH ==Int 0
    rule #getRange(BM, START, WIDTH) => #get(BM, START) +Int (#getRange(BM, START +Int 1, WIDTH -Int 1) *Int 256)
      requires WIDTH >Int 0
```

`#get` looks up a key in a map, defaulting to 0 if the map does not contain the key.
`#set` sets a key in a map, defaulting to 0 if the value is 0.

```k
    syntax Int ::= #get ( Map , Int      ) [function]
    syntax Map ::= #set ( Map , Int, Int ) [function]
 // -------------------------------------------------
    rule #get( (KEY |-> VAL) M, KEY ) => VAL                               [concrete]
    rule #get(               M, KEY ) => 0 requires notBool KEY in_keys(M) [concrete]

    rule #set( M, KEY, VAL ) => M [KEY <- undef] requires VAL ==Int 0  [concrete]

>     rule #set( M, KEY, VAL ) => M [KEY <- VAL  ] requires VAL  >Int 0  [concrete]

Reviewer: Please make these total and mutually exclusive. I'd rather the side-condition be `notBool VAL ==Int 0`, since it makes it obvious that these are total and mutually exclusive.

Of course, the semantics will only ever exercise the `VAL >Int 0` case, but it makes it easier to do reasoning in the Haskell backend when functions are total.

Author: Done
